pkg/controller/state/state_controller.go

```diff
@@ -0,0 +1,223 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+	coreinformersv1 "k8s.io/client-go/informers/core/v1"
+	corev1eventinformers "k8s.io/client-go/informers/core/v1"
+
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+)
+
+const (
+	maxRetries  = 15
+	updateDelay = 5 * time.Second
+
+	// maxUpdateBackoff is the maximum time to react to a change as we back off
+	// in the face of errors.
+	maxUpdateBackoff = 60 * time.Second
+)
+
+type syncFunc struct {
+	name string
+	fn   func(interface{}) error
+}
+
+// need to establish if this is the sort of thing
+// that needs an on/off switch like build controller
+// It probably doesn't, as long as the MCC starts this as
+// it does other controllers
+
+// it seems that other controllers (besides build) do not have a seprate "non generic"
+// controller struct. they just figure out how to do things without data
+// but we want bookkeeping, so we will need to probably have a struct that has a new "bookkeeping" style entity
+// as well as specific health related structs
+type StateControllerConfig struct {
+	UpdateDelay time.Duration
+}
+type StateController interface {
+	Run(int, <-chan struct{}, record.EventRecorder)
+}
+
+type informers struct {
+	operatorCfgInformer operatorcfginformersv1.MachineConfigurationInformer
+	nodeInformer        coreinformersv1.NodeInformer
+	eventInformer       corev1eventinformers.EventInformer
+	serviceInformer     coreinformersv1.ServiceInformer
+}
+type Clients struct {
+	Mcfgclient mcfgclientset.Interface
+	Kubeclient clientset.Interface
+}
+
+type Controller struct {
+	*Clients
+	*informers
+
+	syncHandler   func(key string) error
+	enqueueObject func(interface{})
+
+	serviceInformerSynced cache.InformerSynced
+
+	queue workqueue.RateLimitingInterface
+
+	config StateControllerConfig
+
+	listeners []watch.Interface
+}
+
+func New(
+	serviceInformer coreinformersv1.ServiceInformer,
+	cfg StateControllerConfig,
+	kubeClient clientset.Interface,
+	mcfgClient mcfgclientset.Interface,
+) *Controller {
+
+	ctrl := &Controller{
+		informers: &informers{
+			serviceInformer: serviceInformer,
+		},
+		config: StateControllerConfig{
+			UpdateDelay: time.Second * 5,
+		},
+	}
+
+	ctrl.serviceInformer = serviceInformer
+	ctrl.syncHandler = ctrl.syncStateController
+	ctrl.enqueueObject = ctrl.enqueueDefault
+
+	ctrl.serviceInformerSynced = serviceInformer.Informer().HasSynced
+
+	ctrl.queue = workqueue.NewNamedRateLimitingQueue(workqueue.NewMaxOfRateLimiter(
+		&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(updateDelay), 1)},
+		workqueue.NewItemExponentialFailureRateLimiter(1*time.Second, maxUpdateBackoff)), "machinestatecontroller")
+
+	ctrl.Clients = &Clients{
+		Mcfgclient: mcfgClient,
+		Kubeclient: kubeClient,
+	}
+
+	if ctrl.eventInformer != nil {
+		ctrl.serviceInformer.Informer().AddEventHandler(ctrl.eventHandler())
```

Comment:
lmk if this makes sense. I added this because it basically makes it so that the state controller reacts to any sort of modification of a k8s service. Something needs to trigger the eventHandler... not sure if this is the right thing though